scoville/mvt.py

```diff
@@ -0,0 +1,211 @@
+
+        else:
+            raise ValueError("Unexpected tag %d while decoding value"
+                             % (field.tag,))
+
+    if value is None:
+        raise ValueError("Found no fields when decoding value")
+    if count > 1:
+        raise ValueError("Found multiple fields when decoding value")
+
+    return value
+
+
+class Feature(object):
+
+    def __init__(self, field, keys, values):
+        self.keys = keys
+        self.values = values
+        self.data = field.as_memoryview()
+        self.unpacked = False
+        self.size = field.size
+
+    def __unpack(self):
+        self._fid = None
+        self._tags = []
+        self._geom_type = GeomType.unknown
+        self._cmds_data = []
+        self._tags_size = 0
+        self._cmds_size = 0
+
+        msg = Message(self.data)
+        for field in msg:
+            if field.tag == 1:
+                self._fid = field.as_uint64()
+
+            elif field.tag == 2:
+                subfields = field.as_packed(WireType.varint)
+                self._tags.extend(s.as_uint32() for s in subfields)
+                self._tags_size += field.size
+
+            elif field.tag == 3:
+                self._geom_type = GeomType(field.as_uint32())
+
+            elif field.tag == 4:
+                # don't unpack this now - it's easy enough to iterate over
+                # on-demand.
+                self._cmds_data.append(field)
+                self._cmds_size += field.size
+
+            else:
+                raise ValueError("Unknown Feature tag %d" % field.tag)
+
+        self.unpacked = True
+
+    @property
+    def fid(self):
+        if not self.unpacked:
+            self.__unpack()
+        return self._fid
+
+    @property
+    def geom_type(self):
+        if not self.unpacked:
+            self.__unpack()
+        return self._geom_type
+
+    @property
+    def tags(self):
+        if not self.unpacked:
+            self.__unpack()
+        tags = {}
+        for i in xrange(0, len(self._tags), 2):
+            k = self.keys[self._tags[i]]
+            v = self.values[self._tags[i+1]]
+            tags[k] = _decode_value(v)
+        return tags
+
+    @property
+    def tags_size(self):
+        if not self.unpacked:
+            self.__unpack()
+        return self._tags_size
+
+    @property
+    def cmds_size(self):
+        if not self.unpacked:
+            self.__unpack()
+        return self._cmds_size
+
+
+class Layer(object):
+    """
+    A layer is a container of features, plus some metadata.
+    """
+
+    def __init__(self, field):
+        data = field.as_memoryview()
+
+        self.version = 1
+        self.name = None
+        self.features = []
+        self.keys = []
+        self.values = []
+        self.extent = 4096
+        self.size = field.size
+        self.features_size = 0
+        self.properties_size = 0
+
+        msg = Message(data)
+        for field in msg:
+            if field.tag == 15:
```

Comment:
Would it be helpful to add/use constants for all these checks? The property names you're setting immediately below are explanatory enough for me, but using constants might help with maintainability in the future.

Reply:
Good idea! I've made them all `Enum`s in df42fc6ff841a9145764ac7aac981543d93c6076, so now there's a good place to document them and they're used by name rather than just as magic numbers.

I put them as `Tags` classes internal to the `Feature` and `Layer` classes - does that look right to you?